src/main.cpp

```diff
@@ -2891,6 +2873,48 @@ bool static LoadBlockIndexDB()
     return true;
 }
 
+bool BlockFileIsOpenable(int nFile)
+{
+    CDiskBlockPos pos(nFile, 0);
+    return !CAutoFile(OpenBlockFile(pos, true), SER_DISK, CLIENT_VERSION).IsNull();
+}
+
+bool UndoFileIsOpenable(int nFile)
+{
+    CDiskBlockPos pos(nFile, 0);
+    return !CAutoFile(OpenUndoFile(pos, true), SER_DISK, CLIENT_VERSION).IsNull();
+}
+
+bool DataFilesAreOpenable(int nFile)
+{
+    if (BlockFileIsOpenable(nFile) && UndoFileIsOpenable(nFile))
+        return true;
+    return false;
+}
+
+bool CheckBlockFiles()
+{
+    // Check presence of blk files
+    LogPrintf("Checking all blk files are present...\n");
+    set<int> setRequiredDataFilesAreOpenable;
+    for (CBlockIndex* pindex = chainActive.Tip(); pindex && pindex->pprev; pindex = pindex->pprev) {
+        if (pindex->nStatus & BLOCK_HAVE_DATA && pindex->nStatus & BLOCK_HAVE_UNDO) {
+            if (!setRequiredDataFilesAreOpenable.count(pindex->nFile)) {
```

Comment:
It's not clear to me why you switched from the two-pass solution to a single loop, but from a readability viewpoint I'd prefer something like:
```c++
std::map<int,uint32_t> mapBlkDataFiles;
BOOST_FOREACH(const PAIRTYPE(uint256, CBlockIndex*)& item, mapBlockIndex)
{
    CBlockIndex* pindex = item.second;
    mapBlkDataFiles[pindex->nFile] |= pindex->nStatus;
}
for (std::map<int,uint32_t>::iterator it = mapBlkDataFiles.begin(); it != mapBlkDataFiles.end(); it++)
{
    if ((it->second & BLOCK_HAVE_DATA) && !BlockFileIsOpenable(it->first)) {
        LogPrintf("Error: Required block file %i can't be opened.\n", it->first);
        return false;
    }
    if ((it->second & BLOCK_HAVE_UNDO) && !UndoFileIsOpenable(it->first)) {
        LogPrintf("Error: Required undo file %i can't be opened.\n", it->first);
        return false;
    }
}
```